YunoHost/yunohost
# Config panel actions + applied to domain cert management

## The problem

- We need actions in the config panel for many reasons

## Solution

- Add an action mechanism @_@

## PR Status

- This is mainly a POC for now
- The idea is to have a new component type `button`
- Any section containing a `button` becomes an "action section" which is not displayed by default during `config get`
- `yunohost app config list` lists all the `buttons` ids



## How to test

Example dummy toml snippet in an app :

```toml
[actions]
name = "Actions example"

    [actions.foo]
    name = "Trigger foo action"

        [actions.foo.bar]
        ask = "I can haz bar ?"
        type = "number"
	min = 1
	max = 100
        default = 5

        [actions.foo.force]
        ask = "Force it ?"
        type = "boolean"
        default = false
	visible = "bar > 10"

        [actions.foo.foo]
        ask = "Foo!"
        type = "button"
        icon = "rocket"
        style = "success"
        help = "This should trigger function run__foo"
        enabled = "bar <= 10 or force"  # This doesnt do anything for now but could be useful for condition button (enabled/disabled)
        visible = "bar <= 10 or force"   # This doesnt do anything for now but could be useful for condition button (visible/invisible)

        [actions.foo.foo2]
        ask = "Foo2?"
        type = "button"
        icon = "rocket"
        style = "warning"
        help = "This should trigger function run__foo2"
        enabled = "bar > 50"   # This doesnt do anything for now but could be useful for condition button (enabled/disabled)
        visible = "bar > 50"    # This doesnt do anything for now but could be useful for condition button (visible/invisible)
```

and corresponding controller in `config` script:

```bash
run__foo() {
    ynh_print_info "I am running the action foo"
    touch /tmp/foo_$(date -Iseconds)
}
```

Then run `yunohost app action list config_app` and `yunohost app action run actions.foo.foo` (yeah the app is the full triplet for non-trivial reasons gni)
